src/jit/importer.cpp

```diff
@@ -3918,6 +3918,40 @@ GenTree* Compiler::impIntrinsic(GenTree*                newobjThis,
                 break;
             }
 
+            case NI_BitConverter_ByteSwap:
+            {
+                if (impStackTop(0).val->IsIntegralConst())
+                {
+                    // If the incoming value is a constant, don't bother with the intrinsic step.
+                    // Instead just let morph perform all of the bit shifting and calculate the final value.
+                    isSpecial = true;
+                    break;
+                }
+
+                // The BSWAP instruction is undefined for 16-bit inputs. Instead, use ROR 8 on the
+                // low word register bits. Additionally, BSWAP is undefined for 64-bit inputs on
+                // a 32-bit platform. Instead, fall back to treating the call as non-intrinsic.
+
+                switch (callType)
+                {
+                    case TYP_SHORT:
+                    case TYP_USHORT:
+                        retNode = gtNewOperNode(GT_ROR, callType, impPopStack().val, gtNewIconNode(8));
```

Comment:
That's probably risky business. With some exceptions, the JIT does not use `short`/`ushort` nodes and evaluation is usually done using `int`/`long`.

Reply:
Can you elaborate a bit on the risk? I believe I'm switching on the target method's return type, not the local value node type.

Reply:
You're switching on call type but calls are one of those few nodes that may use small integer types. The other are indirs and lclvar nodes (with indirs being the most common). And those end up producing int values by automatically extending from small int to int.

The net effect is that you may end up with bugs for various reasons:
* ror codegen uses the "actual" node type and emits a 32 bit instruction
* ror codegen use the node type and emits a 16 bit instruction but the consuming node expects a 32 bit value. The upper 16 bits of the 32 bit value may be garbage.
* constant folding may mishandle small int nodes (assuming it pays attention to ror)

I think the biggest danger is that this may appear to work in most cases and fail in some rare and odd circumstances.

Reply:
I was hesitant to add this particular overload in the first place, but it turns out that byte swapping 16-bit values is fairly common in networking stacks. So it seems like this scenario might deserve first-class support in _some_ manner. I used the `ror reg_low16, 8` idiom here because it's the way MSVC++ implements byte swaps over WORD data types. In my experiments, there was always a `movzx reg_32, reg_low16` either immediately before or immediately after the `ror reg_low16, 8` instruction, so that should account for the garbage bits in the high WORD.

But admittedly this is outside my area of expertise, so I'll defer to your judgment here. If you have suggestions for a safer way to do this, I'm all ears. :)

Reply:
> If you have suggestions for a safer way to do this, I'm all ears. :)

Right now, the only idea that comes to mind is to have a `GT_BSWAP16` that is defined as "swaps the lowest 2 bytes of an integer value". Though it's kind of peculiar, I don't think there's any similar case in the JIT today.

cc @CarolEidt Any thoughts one how a 16 bit byte swap could be represented in the JIT's IR? It's somewhat problematic due to JIT's tendency to do everything with 32/64 bit integers.

Reply:
I also think that the right way to deal with the 16-bit case is to use something like `GT_BSWAP16`. The question is whether it should implicitly clear the upper 16 bits, or whether to have the JIT always generate an explicit cast to cause those bits to be cleared. In the latter case, the cast could theoretically be eliminated if the consumer uses only the low bits.